src/main/java/com/powsybl/openloadflow/NetworkCache.java

```diff
@@ -0,0 +1,303 @@
+                }
+            }
+            if (!found) {
+                LOGGER.warn("Cannot update {} of generator '{}'", attribute, generator.getId());
+            }
+            return found;
+        }
+
+        @Override
+        public void onUpdate(Identifiable identifiable, String attribute, String variantId, Object oldValue, Object newValue) {
+            if (contexts == null) {
+                return;
+            }
+            boolean done = false;
+            switch (attribute) {
+                case "v":
+                case "angle":
+                case "p":
+                case "q":
+                case "p1":
+                case "q1":
+                case "p2":
+                case "q2":
+                    // ignore because it is related to state update and won't affect LF calculation
+                    done = true;
+                    break;
+
+                default:
+                    if (identifiable.getType() == IdentifiableType.GENERATOR) {
+                        Generator generator = (Generator) identifiable;
+                        if (attribute.equals("targetV")
+                                && onGeneratorUpdate(generator, attribute, oldValue, newValue)) {
+                            done = true;
+                        }
+                    }
+                    break;
+            }
+
+            if (!done) {
+                reset();
+            }
+        }
+
+        private void onPropertyChange() {
+            // nothing to do there could not have any impact on LF calculation
+        }
+
+        @Override
+        public void onElementAdded(Identifiable identifiable, String attribute, Object newValue) {
+            onPropertyChange();
+        }
+
+        @Override
+        public void onElementReplaced(Identifiable identifiable, String attribute, Object oldValue, Object newValue) {
+            onPropertyChange();
+        }
+
+        @Override
+        public void onElementRemoved(Identifiable identifiable, String attribute, Object oldValue) {
+            onPropertyChange();
+        }
+
+        private void onVariantChange() {
+            // we reset
+            // TODO to study later if we can do better
+            reset();
+        }
+
+        @Override
+        public void onVariantCreated(String sourceVariantId, String targetVariantId) {
+            onVariantChange();
+        }
+
+        @Override
+        public void onVariantOverwritten(String sourceVariantId, String targetVariantId) {
+            onVariantChange();
+        }
+
+        @Override
+        public void onVariantRemoved(String variantId) {
+            onVariantChange();
+        }
+
+        public void close() {
+            reset();
+            if (variantCleaner != null) {
+                variantCleaner.clean();
+                variantCleaner = null;
+            }
+        }
+    }
+
+    private final List<Entry> entries = Collections.synchronizedList(new ArrayList<>());
```

Comment:
Working with a synchronized list allowed you to remove the lock for unitary operations but as mentioned in the javadoc you still need to synchronize the list manually when iterating over it.

Reply:
CopyOnWriteArrayList should be the right one

Reply:
Yes, you can do this with `CopyOnWriteArrayList` to avoid that, but you're not allowed to remove an entry with the iterator then in `evictDeadEntries()`, the unit test `testCacheEviction` fails. You need to do it with `entries.remove(entry)` then.

Reply:
or `entries.removeIf(Predicate<Entry> p)`?